peer/peer.go

```diff
@@ -155,10 +158,66 @@ func IDHexDecode(s string) (ID, error) {
 }
 
 // IDHexEncode returns the hex-encoded multihash representation of the ID.
+//
+// Deprecated: Don't raw-hex encode peer IDs, use base16 CIDs.
 func IDHexEncode(id ID) string {
 	return hex.EncodeToString([]byte(id))
 }
 
+// Decode accepts an encoded peer ID and returns the decoded ID if the input is
+// valid.
+//
+// The encoded peer ID can either be a CID of a key or a raw multihash (identity
+// or sha256-256).
+func Decode(s string) (ID, error) {
+	if strings.HasPrefix(s, "Qm") || strings.HasPrefix(s, "1") {
+		// base58 encoded sha256 or identity multihash
+		m, err := mh.FromB58String(s)
+		if err != nil {
+			return "", fmt.Errorf("failed to parse peer ID: %s", err)
+		}
+		return ID(m), nil
+	}
+
+	c, err := cid.Decode(s)
+	if err != nil {
+		return "", fmt.Errorf("failed to parse peer ID: %s", err)
+	}
+	return FromCid(c)
+}
+
+// Encode encodes a peer ID as a string.
+//
+// At the moment, it base58 encodes the peer ID but, in the future, it will
+// switch to encoding it as a CID by default.
+func Encode(id ID) string {
+	return IDB58Encode(id)
+}
+
+// FromCid converts a CID to a peer ID, if possible.
+func FromCid(c cid.Cid) (ID, error) {
+	ty := c.Type()
+	if ty != cid.Libp2pKey {
+		s := cid.CodecToStr[ty]
+		if s == "" {
+			s = fmt.Sprintf("[unknown multicodec %d]", ty)
+		}
+		return "", fmt.Errorf("can't convert CID of type %s to a peer ID", s)
+	}
+	return ID(c.Hash()), nil
+}
+
+// ToCid encodes a peer ID as a CID of the public key.
+//
+// If the peer ID is invalid (e.g., empty), this will return the empty CID.
+func ToCid(id ID) cid.Cid {
```

Comment:
nit: go style is to capitalise acronyms. We should create an alias `cid.CID => cid.Cid`.

```suggestion
func ToCID(id ID) cid.Cid {
```

Reply:
@raulk  I don't think we should introduce new convention here.
All our codebases already use things like [`NewCidV1`](https://github.com/search?q=org%3Aipfs+NewCidV1&type=Code) and [`FromCid`](https://github.com/search?q=org%3Aipfs+FromCID&type=Code). 

If we choose to change them to use capitalization, that needs to happen across all repos (in separate, coordinated PRs), otherwise we just  introduce noise by mixing conventions.